File: AULA 2/Java/B - Palindromes.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Map<Character, Character> reverseDict = new HashMap<>();
        reverseDict.put('A', 'A');
        reverseDict.put('E', '3'); reverseDict.put('3', 'E');
        reverseDict.put('H', 'H');
        reverseDict.put('I', 'I');
        reverseDict.put('J', 'L'); reverseDict.put('L', 'J');
        reverseDict.put('M', 'M');
        reverseDict.put('O', 'O');
        reverseDict.put('S', '2'); reverseDict.put('2', 'S');
        reverseDict.put('T', 'T');
        reverseDict.put('U', 'U');
        reverseDict.put('V', 'V');
        reverseDict.put('W', 'W');
        reverseDict.put('X', 'X');
        reverseDict.put('Y', 'Y');
        reverseDict.put('Z', '5'); reverseDict.put('5', 'Z');
        reverseDict.put('1', '1');
        reverseDict.put('8', '8');

        Scanner scanner = new Scanner(System.in);
        String result = "";

        while (scanner.hasNext()) {
            
            String word = scanner.next();
            
            boolean isPalindrome = true;
            boolean isMirrored = true;
            List<Character> chars = new ArrayList<>();
            for (char c : word.toCharArray()) {
                chars.add(c);
            }
            int wordLength = chars.size();

            while (!chars.isEmpty()) {
                if (chars.size() >= 2) {
                    char firstChar = shift(chars);
                    char lastChar = pop(chars);

                    isPalindrome = !isPalindrome ? false : firstChar == lastChar;
                    isMirrored = !isMirrored ? false : reverseDict.containsKey(firstChar) && reverseDict.get(firstChar) == lastChar;
                } else if (chars.size() == 1) {
                    char lastChar = pop(chars);
                    isMirrored = !isMirrored ? false : reverseDict.containsKey(lastChar) && reverseDict.get(lastChar) == lastChar;
                }
            }

            result += word
                    + (!isPalindrome && !isMirrored ? " -- is not a palindrome." : "")
                    + (isPalindrome && !isMirrored ? " -- is a regular palindrome." : "")
                    + (!isPalindrome && isMirrored ? " -- is a mirrored string." : "")
                    + (isPalindrome && isMirrored ? " -- is a mirrored palindrome." : "")
                    + "\n\n";
        }

        System.out.print(result);
    }

    public static char pop(List<Character> list) {
        char result = list.get(list.size() - 1);
        list.remove(list.size() - 1);
        return result;
    }

    public static char shift(List<Character> list) {
        char result = list.get(0);
        list.remove(0);
        return result;
    }
}

File: AULA 2/C#/B - Palindromes.cs
using System;
using System.Linq;
using System.Collections.Generic;

public static class Program
{
	public static void Main()
	{
		var reverseDict = new Dictionary<char, char>
		{
			{ 'A', 'A' },
			{ 'E', '3' }, { '3', 'E' },
			{ 'H', 'H' },
			{ 'I', 'I' },
			{ 'J', 'L' }, { 'L', 'J' },
			{ 'M', 'M' },
			{ 'O', 'O' },
			{ 'S', '2' }, { '2', 'S' },
			{ 'T', 'T' },
			{ 'U', 'U' },
			{ 'V', 'V' },
			{ 'W', 'W' },
			{ 'X', 'X' },
			{ 'Y', 'Y' },
			{ 'Z', '5' }, { '5', 'Z' },
			{ '1', '1' },
			{ '8', '8' }
		};

		var result = "";
		
		string word;
		while(!string.IsNullOrEmpty(word = Console.ReadLine()))
		{
			var isPalindrome = true;
			var isMirrored = true;
			var chars = word.ToCharArray().ToList();
			var wordLength = chars.Count();
			
			while(chars.Count > 0)
			{
				if(chars.Count >= 2)
				{
					var firstChar = chars.Shift();
					var lastChar = chars.Pop();

					isPalindrome = !isPalindrome ? false
						: firstChar == lastChar;
					isMirrored = !isMirrored ? false
						: reverseDict.TryGetValue(firstChar, out var mirroredChar) && mirroredChar == lastChar;
				}
				else if(chars.Count == 1)
				{
					var lastChar = chars.Pop();
					isMirrored = !isMirrored ? false
						: reverseDict.TryGetValue(lastChar, out var mirroredChar) && mirroredChar == lastChar;
				}
			}
			
			result += word
				+ (!isPalindrome && !isMirrored ? " -- is not a palindrome." : "")
				+ (isPalindrome && !isMirrored ? " -- is a regular palindrome." : "")
				+ (!isPalindrome && isMirrored ? " -- is a mirrored string." : "")
				+ (isPalindrome && isMirrored ? " -- is a mirrored palindrome." : "")
				+ "\n\n"
			;
		}
		
		Console.Write(result);
	}
	
	public static char Pop(this List<char> list) { var result = list.ElementAt(list.Count - 1); list.RemoveAt(list.Count - 1); return result; }
	public static char Shift(this List<char> list) { var result = list.ElementAt(0); list.RemoveAt(0); return result; }
}